### ISSUE TITLE
[MOD: Base Defense] Чёрное небо и отсутствие падающего снега на карте pve_wintertown

### ISSUE THREAD

**Issue by Vladislav4KZ** [FWGS/xash3d-fwgs#512]
# GoldSource
![Снимок экрана (22).png](https://user-images.githubusercontent.com/22411953/123533778-06f92b80-d73a-11eb-90d1-dfe5ae6909d8.png)

# Xash3D FWGS 0.20.1
![Снимок экрана (21).png](https://user-images.githubusercontent.com/22411953/123533787-1bd5bf00-d73a-11eb-82ca-2a50961c91db.png)

**Comment by Vladislav4KZ** [comment]
В консоли пишет такую ошибку
```
SKY: OpenGL Error: GL_INVALID_ENUM while uploading gfx/env/winterrt [2D]
winterrt, winterbk, winterlf, winterft, winterup, winterdn. done
```
За пределами карты в участке где камера ведёт себя как в воде, этот скайбокс видно

**Comment by a1batross** [comment]
А если с -gldebug запустить движок?

**Comment by Vladislav4KZ** [comment]
> А если с -gldebug запустить движок?

```
[16:36:36]  SKY:   winterrt,  winterbk,  winterlf,  winterft,  winterup,  winterdn. [2021:06:29|16:36:36] done
[2021:06:29|16:36:36] R_GenerateVBOs: allocated array of 31228 verts
[2021:06:29|16:36:36] ^2OpenGL Note:^7 Buffer detailed info: Buffer object 1 (bound to GL_ARRAY_BUFFER_ARB, usage hint is GL_STATIC_DRAW) will use VIDEO memory as the source for buffer object operations.
[2021:06:29|16:36:36] ^2OpenGL Note:^7 Buffer detailed info: Buffer object 2 (bound to GL_ARRAY_BUFFER_ARB, usage hint is GL_STATIC_DRAW) will use VIDEO memory as the source for buffer object operations.
[2021:06:29|16:36:36] ^2OpenGL Note:^7 Buffer detailed info: Buffer object 3 (bound to GL_ARRAY_BUFFER_ARB, usage hint is GL_STATIC_DRAW) will use VIDEO memory as the source for buffer object operations.
[2021:06:29|16:36:36] ^2OpenGL Note:^7 Buffer detailed info: Buffer object 4 (bound to GL_ARRAY_BUFFER_ARB, usage hint is GL_STATIC_DRAW) will use VIDEO memory as the source for buffer object operations.
```

И вот ещё строчки после подключения к серверу
```
[2021:06:29|16:36:37] ^3OpenGL Error:^7 GL_INVALID_ENUM error generated. <mode> is not a valid face culling mode.
[2021:06:29|16:36:37] ^3OpenGL Error:^7 GL_INVALID_ENUM error generated. <mode> is not a valid face culling mode.
[2021:06:29|16:36:37] ^3OpenGL Error:^7 GL_INVALID_ENUM error generated. <mode> is not a valid face culling mode.
[2021:06:29|16:36:37] ^3OpenGL Error:^7 GL_INVALID_ENUM error generated. <mode> is not a valid face culling mode.
```

**Comment by SNMetamorph** [comment]
Нашел в чем причина бага
```
void TriCullFace( TRICULLSTYLE mode )
{
	int glMode;

	switch( mode )
	{
	case TRI_FRONT:
		glMode = GL_FRONT;
		break;
	default:
		glMode = GL_NONE;
		break;
	}

	GL_Cull( mode );
}
```
Нужно поменять `GL_Cull( mode );` на `GL_Cull( glMode );`

**Comment by SNMetamorph** [comment]
Протестил, небо всё так же черное, но снег появился

**Comment by Vladislav4KZ** [comment]
> Протестил, небо всё так же черное, но снег появился

Да, снег появился на него что-то сразу не обратил внимания. И многие спрайты которые раньше не появлялись теперь видны

**Comment by mittorn** [comment]
А что выдаёт при загрузке текстур неба с -gldebug?

**Comment by Vladislav4KZ** [comment]
> А что выдаёт при загрузке текстур неба с -gldebug?

Выше читай, я уже кидал лог с -gldebug

**Comment by Vladislav4KZ** [comment]
> Выше читай, я уже кидал лог с -gldebug
Сейчас даже с -gldebug ничего не пишет, нет ошибок

**Comment by Vladislav4KZ** [comment]
Запустил эту карту как фон меню, через map_background. Небо появилось на секунду и исчезло, стало чёрным

**Comment by nekonomicon** [comment]
Баг еще воспроизводится?

**Comment by Vladislav4KZ** [comment]
> Баг еще воспроизводится?

Да, баг с чёрным небом есть до сих пор.

**Comment by Vladislav4KZ** [comment]
С недавних пор наблюдается такая картина:
Скайбокс теперь отображается нормально, но всё остальное окружение теперь чёрное.

![pve_wintertown_0000.png](https://github.com/FWGS/xash3d-fwgs/assets/22411953/038b7c01-2a44-4cd3-a1ff-d7b855647d98)

Около месяца, или два назад, наблюдался старый баг.

**Comment by Vladislav4KZ** [comment]
В консоли, при запуске этой карты встречается ошибка:

```
Error: No spawn function for info_texlights
```

**Comment by Vladislav4KZ** [comment]
И да, я этот баг перепроверял на ксаше, собранного с master ветки.

В ксаше, собранного с ветки disable-server-unload, естественно то же самое.

**Comment by Vladislav4KZ** [comment]
На других картах вышеописанной проблемы нет.

**Comment by SNMetamorph** [comment]
А как с этой картой обстоят дела на обычном голдсорсе из стима? Может быть поведение отличается от той версии движка которая в Base Defence используется.

**Comment by Vladislav4KZ** [comment]
> А как с этой картой обстоят дела на обычном голдсорсе из стима? Может быть поведение отличается от той версии движка которая в Base Defence используется.
> 

На GoldSource всё нормально.

Я установил Base Defense как обычный мод для Half-Life, и на этой карте есть как и освещение текстур, так и скайбокс отображается нормально, не чёрным цветом.

Я выше говорил про ошибку которая была в консоли ксаша, она также появляется и в GoldSource с включенным режимом разработчика ``developer 2``, только с немного другой формулировкой:

```
Can't init info_texlights
```

Она не влияет на освещение текстур.

**Comment by nekonomicon** [comment]
info_texlights по идее должна удаляться зонеровским компилятором карт, а так эта ентити присутствует в спирите даже не знаю зачем.
По факту наличие ошибки говорит ниочем.

**Comment by a1batross** [comment]
Открыл карту, у меня баг не воспроизводится.

**Comment by a1batross** [comment]
Nevermind, я ошибся.

**Comment by Vladislav4KZ** [comment]
У меня этот мод, как и некоторые другие перестали запускаться после какого-то обновления ксаша.

То же самое было и с оригинальным Half-Life, пока я его не переустановил.

**Comment by Vladislav4KZ** [comment]
Переустановил мод, и он запустился.

Баг остался тот же, про который я писал полгода назад. Там же есть и скриншот как это всё выглядит.

То есть скайбокс отображается нормально, но нет освещения на карте. Все текстуры чёрные (кроме ворот и светильников).

**Comment by a1batross** [comment]
Связано с https://github.com/FWGS/xash3d-fwgs/issues/1368.

Base Defense вызывает pTriApi->Fog в конце HUD_DrawTransparentTriangles с чёрным цветом и очень маленькими значениями flStart/flEnd, что в итоге затемняет всё. 

Если удалить вызовы glFog из TriFog в ref_gl, то проблема пропадает. Как временное решение, могу предложить ввести квар который отключает туман.
